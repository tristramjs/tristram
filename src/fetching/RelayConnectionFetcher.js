/* @flow */
import type { RawSiteMapData } from '../types/sitemap';

import type { Fetcher, FetcherProps } from './index';

require('es6-promise').polyfill();
require('isomorphic-fetch');

type Props<T> = FetcherProps<T> & {
	chunkSize?: number,
	getConnection: GetConnection<T>,
	query: string,
	logErrors?: boolean,
	maxRetries?: number,
};

type GetConnection<T> = <T>(x: Object) => GraphQlConnection<T>;
type GraphQlConnection<T> = { edges: { node: T, cursor: string }[], pageInfo: { hasNextPage: boolean } };

export default class RelayConnection<T> implements Fetcher {
	url: string;
	query: string;
	chunkSize: number;
	transformResult: T => RawSiteMapData;
	logErrors: boolean;
	maxRetries: number;
	getConnection: GetConnection<T>;
	variables: {
		first: number,
		after?: string,
	} = {
		first: 10,
	};

	constructor({
		url, getConnection, transformResult, query, chunkSize, logErrors, maxRetries,
	}: Props<T>) {
		this.url = url;
		this.getConnection = getConnection;
		this.transformResult = transformResult;
		this.query = query;
		this.maxRetries = maxRetries || 3;
		this.chunkSize = chunkSize || 100;
		this.logErrors = logErrors || false;
	}

	async* getData(): AsyncGenerator<RawSiteMapData[], void, void> {
		let hasNextPage;

		do {
			const data: GraphQlConnection<T> = await this.fetchConnection();

			yield data.edges.map(edge => this.transformResult(edge.node));
			({ hasNextPage } = data.pageInfo);

			this.variables = {
				first: this.chunkSize,
				after: data.edges[data.edges.length - 1].cursor,
			};
		} while (hasNextPage);
	}

	async fetchConnection(): Promise<GraphQlConnection<T>> {
		const {
			url, query, variables, getConnection, logErrors,
		} = this;
		let retries = 0;
		const options = {
			method: 'POST',
			headers: {
				'content-type': 'application/json',
			},
			body: JSON.stringify({ query, variables }),
		};

		while (retries < this.maxRetries) {
			const res = await fetch(url, options);
			const data = await res.json();

			if (data.errors && Array.isArray(data.errors)) {
				if (logErrors) {
					// eslint-disable-next-line no-console
					console.log(
						`RelayConnectionFetcher: Received errors in connection:\n"${data.errors
							.map(err => (err.message ? err.message : '<no error message provided>'))
							.join('\n')}"`
					);
				}

				retries = retries + 1;
			} else {
				return getConnection(data);
			}
		}

		throw new Error(`Error fetching connection, tried ${retries} times.`);
	}
}
