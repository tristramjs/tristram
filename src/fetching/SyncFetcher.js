/* @flow */
import type { RawSiteMapData } from '../types/sitemap';

import type { ChunkFetcher } from './index';

export default class SyncFetcher implements ChunkFetcher {
	data: RawSiteMapData | RawSiteMapData[];

	constructor({ data }: { data: RawSiteMapData | RawSiteMapData[] }) {
		this.data = data;
	}

	async* getData(): AsyncIterator<RawSiteMapData[]> {
		if (Array.isArray(this.data)) {
			yield this.data;
		} else {
			yield [ this.data ];
		}
	}
}
