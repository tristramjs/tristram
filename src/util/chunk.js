/* @flow */

function chunk<T>(arr: T[], chunkSize: number): T[][] {
	const chunks = [];
	let index = 0;
	const { length } = arr;

	while (index < length) {
		chunks.push(arr.slice(index, (index += chunkSize)));
	}

	return chunks;
}

export default chunk;
