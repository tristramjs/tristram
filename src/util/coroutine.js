/* @flow */

/**
 * Returns a function that, when called,
 * returns a generator object that is immediately
 * ready for input via `next()`
 */
export default function coroutine(generatorFunction: () => AsyncGenerator<*, *, *>) {
	return function cb(...args: any) {
		const generatorObject = generatorFunction(...args);
		generatorObject.next();
		return generatorObject;
	};
}
