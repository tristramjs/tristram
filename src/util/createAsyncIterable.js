/* @flow */

export default async function* createAsyncIterable(syncIterable: any): AsyncGenerator<*, *, *> {
	for (const elem of syncIterable) {
		yield elem;
	}
}
