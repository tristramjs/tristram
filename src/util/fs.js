/* @flow */
import fs from 'fs';
import util from 'util';

const appendFile = util.promisify(fs.appendFile);
const exists = util.promisify(fs.exists);
const mkdir = util.promisify(fs.mkdir);
const readdir = util.promisify(fs.readdir);
const readfile = util.promisify(fs.readFile);
const unlink = util.promisify(fs.unlink);
const rmdir = util.promisify(fs.rmdir);
const writeFile = util.promisify(fs.writeFile);

export { appendFile, exists, mkdir, readdir, readfile, unlink, rmdir, writeFile };
