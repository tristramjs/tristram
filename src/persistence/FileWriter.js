/* @flow */
import { appendFile, writeFile } from '../util/fs';

import appendToXml from './appendToXml';

import type { Writer } from './index';

type Props = {
	path: string,
	fileName: string,
};

export default class FileWriter implements Writer {
	path: string;
	fileName: string;
	sitemaps: number;
	sitemap: AsyncGenerator<void, void, string>;

	constructor({ path, fileName }: Props) {
		this.path = path;
		this.fileName = fileName;
		this.sitemaps = 0;
	}

	async createSitemap(): Promise<string> {
		const path = this.getSitemapPath();
		await writeFile(path, '');
		this.sitemap = appendToXml(data => appendFile(path, data));
		this.sitemaps = this.sitemaps + 1;

		return path;
	}

	async writeChunk(data: string) {
		if (this.sitemap) {
			await this.sitemap.next(data);
		} else {
			throw new Error('Cant write to file. Did you forget to call/await `createSitemap`?');
		}
	}

	async commitSitemap() {
		if (this.sitemap) {
			await this.sitemap.return();
		} else {
			throw new Error('Cant write to file. Did you forget to call/await `createSitemap`?');
		}
	}

	async createIndexSitemap(data: string) {
		await writeFile(`${this.path}indexSitemap.xml`, data);
	}

	getSitemapPath() {
		return `${this.path}${this.fileName}-${this.sitemaps}.xml`;
	}

	getPath() {
		return this.path;
	}
}
