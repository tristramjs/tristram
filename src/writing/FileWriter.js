/* @flow */
import { appendFile, writeFile } from '../util/fs';

import type { Writer, Path } from './index';

type Props = {
	path: Path,
};

export default class FileWriter implements Writer {
	path: Path;
	fileName: string;

	constructor({ path }: Props) {
		this.path = path;
	}

	async write(key: string, data: string) {
		const path = this.path + key;
		await writeFile(path, data);
		return path;
	}
}
