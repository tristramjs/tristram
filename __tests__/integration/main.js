/* @flow */

import { setup, cleanup } from '../__testHelpers__/fs';
import { readdir, readfile } from '../../src/util/fs';
import Main from '../../src/main';
import SyncFetcher from '../../src/fetching/SyncFetcher';
import RelayConnectionFetcher from '../../src/fetching/RelayConnection';
import FileWriter from '../../src/persistence/FileWriter';
import PlainFormatter from '../../src/formatting/Plain';

const path = `${process.cwd()}/MainIntegrationTest/`;

describe('Main module', () => {
	beforeAll(setup(path));

	it('should run without errors', async () => {
		const options = { hostname: 'http://foo.bar', maxItemsPerSitemap: 2 };

		const syncFetcher = new SyncFetcher({
			data: [ { loc: 'quz' }, { loc: 'qaz' }, { loc: 'qak' } ],
		});

		const chunkFetcher = new RelayConnectionFetcher({
			url: 'http://test.com/graphql',
			query: /* GraphQL */ `query Test($first: Int!, $after: String) {
				viewer {
					allTestItems(first: $first, after: $after) {
						pageInfo {
							hasNext
						}
						edges {
							cursor
							node {
								id
							}
						}
					}
				}
			}`,
			getConnection: conn => conn.data.viewer.allTestItems,
			transformResult: (data) => {
				const returnData = { ...data };
				returnData.loc = data.id;
				delete returnData.id;
				if (data.lastmod) {
					returnData.lastmod = new Date(data.lastmod);
				}
				if (data.video) {
					returnData.video = data.video.map((item) => {
						const returnItem = { ...item };
						if (item.expiration_date) {
							returnItem.expiration_date = new Date(item.expiration_date);
						}
						if (item.publication_date) {
							returnItem.publication_date = new Date(item.publication_date);
						}
						return returnItem;
					});
				}
				return returnData;
			},
		});

		const main = new Main({
			fetchers: [ syncFetcher, chunkFetcher ],
			formatter: new PlainFormatter(),
			writer: new FileWriter({ path, fileName: 'sitemap' }),
			options,
		});

		await main.run();

		const files = await readdir(path);
		for (const file of files) {
			expect(await readfile(`${path}/${file}`, 'utf8')).toMatchSnapshot();
		}

		expect(files.length).toBe(4);
	});

	afterAll(cleanup(path));
});
