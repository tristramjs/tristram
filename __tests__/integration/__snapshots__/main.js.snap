// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main module should run without errors 1`] = `"<foo><node loc=\\"quz\\"/><node loc=\\"qaz\\"/></foo>"`;

exports[`Main module should run without errors 2`] = `"<foo><node loc=\\"qak\\"/><node loc=\\"bla\\"/></foo>"`;

exports[`Main module should run without errors 3`] = `"<foo><node loc=\\"blerb\\"/></foo>"`;

exports[`Main module should run without errors 4`] = `"<?xml version=\\"1.0\\"?><urlset xmlns=\\"http://www.sitemaps.org/schemas/sitemap/0.9\\" xmlns:image=\\"http://www.google.com/schemas/sitemap-image/1.1\\" xmlns:video=\\"http://www.google.com/schemas/sitemap-video/1.1\\"><loc>three</loc><loc>four</loc></urlset>"`;
